anvio/synteny.py

```diff
@@ -0,0 +1,303 @@
+
+
+class NGram(object):
+    """class for counting NGrams
+
+    anvi-analyze-synteny is designed to work with a group of similar loci, where each locus is a contig (which
+    can lie in any number of contigs dbs)
+
+    Parameters
+    ==========
+    args : argparse.Namespace
+        For examples, arguments accepted by anvi-analyze-syntenty
+
+    skip_sanity_check : bool, False
+        If True, sanity_check will not be called.
+
+    Notes
+    =====
+    - Currently the design assumes that each locus is a contig. In the future we have plans to expand this to
+      compare genomes, not just loci, to each other. If that behavior is desired in the current design, each
+      genome should be a single contig.
+    """
+
+    def __init__(self, args, run=terminal.Run(), progress=terminal.Progress(), skip_sanity_check=False):
+        """Parses arguments and run sanity_check"""
+
+        self.args = args
+        self.run = run
+        self.progress = progress
+
+        # Parse arguments
+        A = lambda x: args.__dict__[x] if x in args.__dict__ else None
+        self.external_genomes = A('external_genomes')
+        self.annotation_source = A('annotation_source')
+        self.window_range = A('window_range')
+        self.in_in_unknowns_mode = A('analyze_unknown_functions')
+        self.external_genomes = utils.get_TAB_delimited_file_as_dictionary(self.external_genomes)
+        self.output_file = A('output_file')
+
+        # This houses the ngrams' data
+        self.ngram_attributes_list = []
+
+        self.num_contigs_in_external_genomes_with_genes = 0
+
+        if not skip_sanity_check:
+            self.sanity_check()
+
+        # unless we are in debug mode, let's keep things quiet.
+        if anvio.DEBUG:
+            self.run_object = terminal.Run()
+        else:
+            self.run_object = terminal.Run(verbose=False)
+
+
+    def sanity_check(self):
+        """Sanity_check will confirm input for NGram class"""
+
+        # checking if the annotation source is common accross all contigs databases
+        g = genomedescriptions.GenomeDescriptions(self.args)
+        g.load_genomes_descriptions(init=False)
+
+        if self.annotation_source not in g.function_annotation_sources:
+            raise ConfigError("The annotation source you requested does not appear to be in the "
+                              "contigs database :/ "
+                              "Please confirm you are only providing one annotation source :)")
+
+        if not self.args.output_file:
+            raise ConfigError("You should provide an output file name.")
+
+        # checking window-range input
+        if ":" not in self.window_range:
+            raise ConfigError("anvi'o would love to slice and dice your loci, but... the "
+                              "Format of window_range must be x:y (e.g. Window sizes 2 to 4 would be denoted as: 2:4)")
+
+        self.window_range = [int(n) for n in self.window_range.split(":")]
```

Comment:
What happens when you send `a:b` as window range? :) Probably `int(n)` will explode, right?

So this line could have been in a try/except block to avoid that, and if there is a `TypeError`, the user could be warned to use only integers.

Reply:
Yup just tested this.

What do you think of this solution:

```
self.window_range = [n for n in self.window_range.split(":")]

if not isinstance(self.window_range[0], int) or not isinstance(self.window_range[1], int):
        raise ConfigError("anvi'o would love to slice and dice your loci, but the "
            "window-ranges need to be integers :)")
```

Reply:
nvm that solution didnt work. I understand what your saying about try/except

How's this:

```
try:
    self.window_range = [int(n) for n in self.window_range.split(":")]
except:
    raise ConfigError("anvi'o would love to slice and dice your loci, but the "
                              "window-ranges need to be integers :)")
```

Reply:
Great. But the best practice is this: You are trying to catch an error due to conversion of a value by the 'int' function. So this is what I do:

``` python
$ python
Python 2.7.10 (default, Aug 17 2018, 19:45:58)
[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.0.42)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> int('a')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'a'
>>>
```

So this tells me that I should `expect` a `ValueError` if someone sends a string instead of an int. This way if there is something else is wrong, it would be reported full, rather than saying "your window is not integer". Makes sense?

(so basically I'm saying change `expect:` to `expect ValueError:`).

Reply:
Gotcha, I had no idea one could be that specific with errors!